### ISSUE TITLE
Workstation LogIn Updater

### ISSUE THREAD

**Issue by ninavizz** [freedomofpress/securedrop-ux#98]
## Problem
For reasons, updates have to happen before a user engages with Qubes/SD tasks on the Workstation... and, when the user cannot close their laptop or cancel the updates. Should a user cancel updates mid-process, it could potentially break the Workstation to the point of needing hours of expert fixes. 

## Solution
Mickael & Jen already developed [this swanky solution](https://github.com/freedomofpress/securedrop-workstation/pull/396), and it just needs some UX luv.

**Comment by emkll** [comment]
Here are some screenshots from the latest version of the script, the text could definitely use some work. This should cover all the user-facing scenarios.

#### Once launched, the application search for updates:
![checking_for_updates](https://user-images.githubusercontent.com/15223328/72192632-722f3080-33d4-11ea-98eb-c05d3816032f.png)

#### If the system is up-to-date (no updates need to be applied), the user may start the client:
![system_already_up_to_date](https://user-images.githubusercontent.com/15223328/72192647-81ae7980-33d4-11ea-9273-330e63dd2261.png)

#### If updates are required, the user is asked if they want to apply VM updates:
![updates_required](https://user-images.githubusercontent.com/15223328/72192710-8ffc9580-33d4-11ea-90aa-90642a7037cc.png)

#### Once the user clicks "Start Updates", updates are being applied to the VMs
![updates_in_progress](https://user-images.githubusercontent.com/15223328/72192729-9f7bde80-33d4-11ea-9213-1e1b34d41e15.png)

#### If the updates require a reboot of the entire workstation, the user is presented with a reboot button:
![reboot_required](https://user-images.githubusercontent.com/15223328/72192748-ad316400-33d4-11ea-8927-720d5ed6db1b.png)

#### If the updates do not require a full workstation reboot, the user is offered the option top open the client:
![updates_applied](https://user-images.githubusercontent.com/15223328/72192766-cb975f80-33d4-11ea-9fce-c51e0c4b70fd.png)

**Comment by ninavizz** [comment]
U da bomb, thx Mickael! :)

**Comment by ninavizz** [comment]
First 2 iterations, [shown in this PDF.](https://drive.google.com/open?id=1UyeY3J0lXQZS69XGo-6dR0INFnHaNMH4) 

## Agreed-to MoSCoW for Beta, based on 2nd iteration:
- must: rename tool to `Workstation preflight updater`
- must: make errors/warnings very obvious
- must: tweak text, ideally as rich-text that we can lay out similar to your screens
- should: change three button flow as shown in your screens
- could: include icons w/in top bars (will include in a follow-up issue)
- could: change progress bar styling

![image](https://user-images.githubusercontent.com/8262612/72478744-8e054e80-37a7-11ea-9e77-88f9c3db929d.png)

**Comment by emkll** [comment]
Version 0.2, changed the wording and the layout:

Note that the buttons are always in the same location (they are just hidden/shown), centering them may be somewhat more complex to implement.

Once launched, the application search for updates:

![checking_for_updates](https://user-images.githubusercontent.com/15223328/72547739-32949880-385b-11ea-830a-654cc440eedc.png)

If the system is up-to-date (no updates need to be applied), the user may start the client:

![no_updates_today](https://user-images.githubusercontent.com/15223328/72547741-36c0b600-385b-11ea-92db-50b0bc67cbf3.png)

If updates are required, the user is asked if they want to apply VM updates:

![updates_required](https://user-images.githubusercontent.com/15223328/72547762-3b856a00-385b-11ea-8fc8-8d66b66892b1.png)

Once the user clicks "Start Updates", updates are being applied to the VMs

![updates_in_progress](https://user-images.githubusercontent.com/15223328/72547781-46d89580-385b-11ea-9e39-ad4aa2753a5d.png)

If the updates require a reboot of the entire workstation, the user is presented with a reboot button:

![updates_complete_reboot_required](https://user-images.githubusercontent.com/15223328/72547790-4dffa380-385b-11ea-868a-71e61068eacb.png)

If the updates do not require a full workstation reboot, the user is offered the option top open the client:
![workstation_up_to_date](https://user-images.githubusercontent.com/15223328/72547799-535cee00-385b-11ea-9be4-4147d9202a34.png)

**Comment by ninavizz** [comment]
@emkll if the user were to close the VM window while it's _checking_ for updates, would that cause probs for the next person to try launching the app?

**Comment by emkll** [comment]
> @emkll if the user were to close the VM window while it's checking for updates, would that cause probs for the next person to try launching the app?

Closing the updater window while it is _checking_ for updates (not _applying_) does not pose a problem

**Comment by ninavizz** [comment]
![image](https://user-images.githubusercontent.com/8262612/72558882-720db580-3858-11ea-869c-bc955b3fc937.png)

Ok! Complete spec for updates as discussed in today's UX meeting, [is here](https://drive.google.com/open?id=1llp5lX9AsPvYl0S-6Q1aO76hdedHqga9). For the most part, no text edits are included in the above—the warning screen, being the lone exception.

Where there may be scope-creep from what we've discussed, I tried to clearly denote. @redshiftzero and @eloquence you guys' call on who is allowed to do what, beyond Mickael and Allie's agreed-to tasks in the meeting... as I appreciate each's time is being copiously budgeted by the minute, at this stage in our Beta prep. :)

**Comment by ninavizz** [comment]
...also of note: I reversed course on my agreement to do right-aligned buttons, per David's flag. Because the buttons are so wide as they are sized right now, it looked "off" to have them right aligned—SO for now, I wanted to keep them all center-aligned. If that's cool with others.

**Comment by emkll** [comment]
Version 0.3, I think this is the best we can do without major changes to the UI code. Other than string changes, I propose we address any further changes in more specific follow-ups tasks (likely post-beta).

#### Once launched, the application search for updates:

![checking_for_updates](https://user-images.githubusercontent.com/15223328/72647334-4e289d80-3946-11ea-9772-b8de711ddbe7.png)


#### If the system is up-to-date (no updates need to be applied), the user may start the client:

![no_updates_today](https://user-images.githubusercontent.com/15223328/72647347-554fab80-3946-11ea-8a42-312e8fbb3aa4.png)


#### If updates are required, the user is asked if they want to apply VM updates:
![security_updates_required](https://user-images.githubusercontent.com/15223328/72647351-5b458c80-3946-11ea-9215-d1adc2415f63.png)


#### Once the user clicks "Start Updates", updates are being applied to the VMs

![updates_in_progress](https://user-images.githubusercontent.com/15223328/72647354-600a4080-3946-11ea-8d07-81b77bcd569d.png)


#### If the updates require a reboot of the entire workstation, the user is presented with a reboot button:

![updates_complete_reboot_required](https://user-images.githubusercontent.com/15223328/72647361-65678b00-3946-11ea-80bc-0ce56d3e44d3.png)


#### If the updates do not require a full workstation reboot, the user is offered the option top open the client:

![updates_applied](https://user-images.githubusercontent.com/15223328/72647365-6a2c3f00-3946-11ea-977e-799e5faf3dce.png)

#### If an error occurred during the update process, the user is informed that updates have failed:
![error_applying_updates](https://user-images.githubusercontent.com/15223328/72647394-7ca67880-3946-11ea-83a8-0a6eb9cf4383.png)

**Comment by ninavizz** [comment]
For Beta, I believe this issue has been sufficiently addressed and can be considered kuh-lozed.

**Comment by ninavizz** [comment]
Question @emkll: does the text read `reboot may be required` vs two different strings to definitively state yay or nay wrt updates, because that'd be a heavy code lift? Assuming "yes;" if yes, I will create a new ticket to add that screen post-beta. :)

**Comment by emkll** [comment]
Unfortunately a yes due to how we handle the reboot logic for users that decide to click "cancel" after being prompted for a reboot. The complexity here is that we want to make sure they are prompted to reboot, even when when they re-open the updater after applying updates.
